contributors/design-proposals/admission_control_event_rate_limit.md

```diff
@@ -0,0 +1,149 @@
+# Admission control plugin: EventRateLimit
+
+## Background
+
+This document proposes a system for using admission control to enforce a limit
+on the number of event requests that the API Server will accept in a given time
+slice. In a large multi-tenant cluster, there may be a small percentage of
+tenants that are always in some type of error state, in which each tenant is
+producing a steady stream of error event requests. Each individual tenant may not
+be producing an absurd amount of event requests on its own, but taken collectively
+the errors from this small percentage of tenants can have a significant impact on
+the performance of the cluster overall. 
+
+## Use cases
+
+1. Ability to protect the API Server from being flooded by event requests.
+2. Ability to protect the API Server from being flooded by event requests for
+   a particular namespace.
+3. Ability to protect the API Server from being flooded by event requests for
+   a particular user.
+4. Ability to protect the API Server from being flooded by event requests from
+   a particular source+object.
+
+## Data Model
+
+### Configuration
+
+```go
+// Configuration provides configuration for the EventRateLimit admission controller.
+type Configuration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Limits to place on events received.
+	// Limits can be placed on events received server-wide, per namespace,
+	// per user, and per source+object.
+	// At least one limit is required.
+	Limits []Limit `json:"limits"`
+}
+
+type Limit struct {
+	// Type of limit.
+	// The following are valid values.
+	// "server": limits are maintained against all events received by the server
+	// "namespace": limits are maintained against events from each namespace
+	// "user": limits are maintained against events from each user
+	// "source+object": limits are maintained against events from each source+object
+	Type string `json:"type"`
+
+	// Maximum QPS of events for this limit
+	QPS float32 `json:"qps"`
+
+	// Maximum burst for throttle of events for this limit
+	Burst int ` json:"burst"`
+
+	// Maximum number of limits to maintain. If room is needed in the cache for a
+	// new limit, then the least-recently used limit is evicted, resetting the
+	// stats for that subset of the universe.
+	//
+	// For example, if the type of limit is "namespace" and the limit for
+	// namespace "A" is evicted, then the next event received from namespace "A"
+	// will use reset stats, enabling events from namespace "A" as though no
+	// events from namespace "A" have yet been received.
+	//
+	// If the type of limit is "server", then CacheSize is ignored and can be
+	// omitted.
+	CacheSize int `json:"cacheSize"`
```

Comment:
Comment is hard to parse for me (though I think I finally understood it)

Is this really needed? What kind of cache it is? LRU? Generally it makes the semantics rather strange. E.g. when you have limit, say 5, and 7 namespaces in the system and you get Events from all namespaces regularly you'll end up with small number of used limit for all namespaces. I think if we want to implement this, we need to store information about all "buckets". Source+object is most problematic in this case, but this can be handled client-side, so it should be fine to remove it.

Reply:
> I think if we want to implement this, we need to store information about all "buckets". Source+object is most problematic in this case, but this can be handled client-side, so it should be fine to remove it.

Turns out you don't really need to store it to have a big impact.  Since it ramps up fairly quickly, keeping the information locally, per-apiserver, is enough to prevent crushing an apiserver.

Reply:
Sorry, I wrote it in a confusing way. I don't mean storing in etcd or anything, I mean by "storing" in cache. I.e. limited number of entries in cache is problematic in this case IMO.

Reply:
The cache is used to place an upper-bound on the memory usage of the admission controller. The cache size can be set large enough to fit accommodate all of the "buckets" that exist when the system is in a steady state.

For example, if your system has 7 namespaces, you would use a cache size for "namespace" that is at least 7. If you had the admission controller configured poorly, so that the cache size for "namespace" is only 5, the admission controller is still likely to serve its purpose, assuming that you don't have more than 5 namespaces behaving poorly, where poorly behaving means sending too many events to the API Server. The poorly behaving namespaces are least likely to be evicted from the cache, by the very nature of their poor behavior.

If we did not use a cache, then a bad actor could bring down the API Server over time being simply creating and deleting namespaces. The downside of having unbounded memory usage seems to me to be far greater than occasionally having a "bucket" be granted a greater number of event requests than it should.

Reply:
But we can watch namespaces and delete non-existing ones ourselves, right? If you expect to have a bucket per "class" then I think there's even less reason to limit size of the cache (as you assume it's practically unbounded). This would add a place when cluster admin can configure something wrong, and given the number of kubernetes clusters out there someone will and this is not something I'd want to debug.

Reply:
Pardon my ignorance as I am somewhat new to kubernetes. What does "class" mean here? I am having a hard time following the logic of why assuming that something is practically unbounded would make you less inclined to set up a cache for it. I would think that the expectation that something is unbounded would be exactly the reason why you would want/need to place limits on it with something like a cache. Are you suggesting that we could watch and delete buckets for source+object and user as well, and that that would be a practically way of limiting those?

I am also struggling to understand what you expect the catastrophe to be of a poorly configured cache size. If the cache size is configured too large, then that is no worse than having no cache. If the cache size if configured too small, then the admission control may allow some more events through than it should. Is there some other thing that may happen that I am not seeing? This admission control is a last effort to protect the API Server from being overwhelmed by events, but is not the only protection in place. There is already event rate limiting client-side, assuming that the client is using the kubernetes event recorder.

Reply:
By class I meant abstraction class (e.g. all events from the same namespace). I wasn't clear what I meant. Cache itself is a good idea. I don't like it being bounded (hence the need for this field).

As I wrote I believe source+object limiting can be done client side (as source is defined by client), so I think we can just drop it here.

Catastrophe is a big word - I expect people to be confused. They set up admission controller to rate limit their Event influx, and probably use default values for cache sizes because that's what you do when you start. Then they forget about this setting and everything works until it doesn't. They'll see that more Events getting into the API server that it should and they'll try to figure out why, which I don't expect to be easy.

Generally I think that bounding the cache is causing more harm than good.